### ISSUE TITLE
Allow where clause in subgraph queries to reference another field in type

### ISSUE THREAD

**Issue by cmwhited** [graphprotocol/graph-node#4332]
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

Currently, the `where` clause in a subgraph query only allows you to pass values to it that are of the type of the field being filtered against.

ex:
```graphql
query Indexers {
  indexers(where: { stakedTokens: 0 }) {
    id
    stakedTokens
    lockedTokens
  }
}
```
Where: `stakedTokens` is of type `BigInt` and the value passed (0 in the example) must also be a `BigInt`.

**What is the expected behavior?**

It would be very helpful if the `where` clause was able to compare two fields against each other to filter results.

ex:
```graphql
query Indexers {
  indexers(where: { stakedTokens: {indexer? me? this?}.lockedTokens }) {
    id
    stakedTokens
    lockedTokens
  }
}
```

Mapping down to sql, this would look (something like):
```sql
SELECT id, staked_tokens, locked_tokens FROM indexer WHERE staked_tokens = locked_tokens;
```

**Comment by llllvvuu** [comment]
It would also be very helpful if the RHS could be a nested value, like:
```
purchase(where: { timestamp_gte: this.vault.epochStart }) {
  id
  timestamp
  vault {
    id
    epochStart    
  }
}
```

looking something like
```
SELECT purchase.id as purchase_id, timestamp, vault.id as vault_id, epoch_start
FROM purchase JOIN vault
WHERE purchase.timestamp >= vault.epoch_start
```

**Comment by saihaj** [comment]
This wouldn't be valid GraphQL syntax which makes it really tricky to implement

**Comment by llllvvuu** [comment]
How about generated columns? https://stackoverflow.com/questions/61674263/compare-two-fields-in-graphql-hasura#comment109109862_61680466

`field1_gt_field2`

**Comment by azf20** [comment]
hey @cmwhited this makes a lot of sense from a user perspective.
@llllvvuu interesting proposal on generated columns - I think a  challenge with the automated code generation to cover all cases would be an explosion of filters in the schema.
@saihaj in terms of valid graphQL syntax, is any reference of another field invalid (not just nested?)
I wonder if this is a focused use-case for [custom resolvers](https://github.com/graphprotocol/graph-node/issues/880)

**Comment by saihaj** [comment]
> @saihaj in terms of valid graphQL syntax, is any reference of another field invalid (not just nested?)
@azf20 Yes. It is just not possible since the GraphQL specification makes resolutions of queries parallel whenever possible. Having a field dependent on a field would make this harder and we would build a custom execution engine to support. 

Short term way we can try of thinking a way in `schema.graphql` file so users can annotate these are fields they want to compare against in query so we generate a very specific query.

Long term way is figuring is [custom resolvers](https://github.com/graphprotocol/graph-node/issues/880)
